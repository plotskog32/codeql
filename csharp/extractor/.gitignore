obj/
SemmleTests/bin
TestResults/
*.manifest
*.pdb
*.suo
*.mdb
*.vsmdi
CSharpExtractor*.config
csharp.log
**/bin/Debug
**/bin/Release
*.tlog
.vs
packages
*.user