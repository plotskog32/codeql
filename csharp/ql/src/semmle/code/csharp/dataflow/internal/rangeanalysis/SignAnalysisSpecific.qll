/**
 * Provides C#-specific definitions for use in sign analysis.
 */
module Private {
  private import SsaUtils as SU
  private import csharp as CS
  private import ConstantUtils as CU
  private import semmle.code.csharp.controlflow.Guards as G
  import Impl

  class Guard = G::Guard;

  class ConstantIntegerExpr = CU::ConstantIntegerExpr;

  class SsaVariable = CS::Ssa::Definition;

  class SsaPhiNode = CS::Ssa::PhiNode;

  class VarAccess = CS::AssignableAccess;

  class FieldAccess = CS::FieldAccess;

  class CharacterLiteral = CS::CharLiteral;

  class IntegerLiteral = CS::IntegerLiteral;

  class LongLiteral = CS::LongLiteral;

  class CastExpr = CS::CastExpr;

  class Type = CS::Type;

  class Expr = CS::Expr;

  class VariableUpdate = CS::AssignableDefinition;

  class ExprWithPossibleValue = CS::Expr;

  class Field = CS::Field;

  predicate ssaRead = SU::ssaRead/2;
}

private module Impl {
  private import csharp
  private import SsaUtils
  private import ConstantUtils
  private import semmle.code.csharp.controlflow.Guards
  private import Linq.Helpers
  private import Sign
  private import SignAnalysisCommon
  private import SsaReadPositionCommon
  private import semmle.code.csharp.commons.ComparisonTest

  private class BooleanValue = AbstractValues::BooleanValue;

  /** Gets the character value of expression `e`. */
  string getCharValue(Expr e) { result = e.getValue() and e.getType() instanceof CharType }

  /**
   * Holds if `e` is an access to the size of a container (`string`, `Array`,
   * `IEnumerable`, or `ICollection`).
   */
  predicate containerSizeAccess(Expr e) {
    exists(Property p | p = e.(PropertyAccess).getTarget() |
      propertyOverrides(p, "System.Collections.Generic.IEnumerable<>", "Count") or
      propertyOverrides(p, "System.Collections.ICollection", "Count") or
      propertyOverrides(p, "System.String", "Length") or
      propertyOverrides(p, "System.Array", "Length")
    )
    or
    e instanceof CountCall
  }

  /** Holds if `e` is by definition strictly positive. */
  predicate positiveExpression(Expr e) { e instanceof SizeofExpr }

  abstract class NumericOrCharType extends Type { }

  class UnsignedNumericType extends NumericOrCharType {
    UnsignedNumericType() {
      this instanceof CharType
      or
      this instanceof UnsignedIntegralType
      or
      this instanceof PointerType
      or
      this instanceof Enum and this.(Enum).getUnderlyingType() instanceof UnsignedIntegralType
    }
  }

  class SignedNumericType extends NumericOrCharType {
    SignedNumericType() {
      this instanceof SignedIntegralType
      or
      this instanceof FloatingPointType
      or
      this instanceof DecimalType
      or
      this instanceof Enum and this.(Enum).getUnderlyingType() instanceof SignedIntegralType
    }
  }

  /** Returns the underlying variable update of the explicit SSA variable `v`. */
  AssignableDefinition getExplicitSsaAssignment(Ssa::ExplicitDefinition v) {
    result = v.getADefinition()
  }

  /** Returns the assignment of the variable update `def`. */
  Expr getExprFromSsaAssignment(AssignableDefinition def) { result = def.getSource() }

  /** Holds if `def` can have any sign. */
  predicate explicitSsaDefWithAnySign(AssignableDefinition def) {
    not exists(def.getSource()) and
    not def.getElement() instanceof MutatorOperation
  }

  /** Returns the operand of the operation if `def` is a decrement. */
  Expr getDecrementOperand(AssignableDefinition def) {
    result = def.getElement().(DecrementOperation).getOperand()
  }

  /** Returns the operand of the operation if `def` is an increment. */
  Expr getIncrementOperand(AssignableDefinition def) {
    result = def.getElement().(IncrementOperation).getOperand()
  }

  /** Gets the variable underlying the implicit SSA variable `v`. */
  Declaration getImplicitSsaDeclaration(Ssa::ImplicitDefinition v) {
    result = v.getSourceVariable().getAssignable()
  }

  /** Holds if the variable underlying the implicit SSA variable `v` is not a field. */
  predicate nonFieldImplicitSsaDefinition(Ssa::ImplicitDefinition v) {
    not getImplicitSsaDeclaration(v) instanceof Field
  }

  /** Returned an expression that is assigned to `f`. */
  Expr getAssignedValueToField(Field f) {
    result = f.getAnAssignedValue() or
    result = any(AssignOperation a | a.getLValue() = f.getAnAccess())
  }

  /** Holds if `f` can have any sign. */
  predicate fieldWithUnknownSign(Field f) { not f.fromSource() or not f.isEffectivelyPrivate() }

  /** Holds if `f` is accessed in an increment operation. */
  predicate fieldIncrementOperationOperand(Field f) {
    any(IncrementOperation inc).getOperand() = f.getAnAccess()
  }

  /** Holds if `f` is accessed in a decrement operation. */
  predicate fieldDecrementOperationOperand(Field f) {
    any(DecrementOperation dec).getOperand() = f.getAnAccess()
  }

  /** Returns possible signs of `f` based on the declaration. */
  pragma[inline]
  Sign specificFieldSign(Field f) { not exists(f.getInitializer()) and result = TZero() }

  /**
   * Holds if `e` has type `NumericOrCharType`, but the sign of `e` is unknown.
   */
  predicate numericExprWithUnknownSign(Expr e) {
    e.getType() instanceof NumericOrCharType and
    not e = getARead(_) and
    not e instanceof FieldAccess and
    not e instanceof TypeAccess and
    // The expression types that are listed here are the ones handled in `specificSubExprSign`.
    // Keep them in sync.
    not e instanceof AssignExpr and
    not e instanceof AssignOperation and
    not e instanceof UnaryPlusExpr and
    not e instanceof PostIncrExpr and
    not e instanceof PostDecrExpr and
    not e instanceof PreIncrExpr and
    not e instanceof PreDecrExpr and
    not e instanceof UnaryMinusExpr and
    not e instanceof ComplementExpr and
    not e instanceof AddExpr and
    not e instanceof SubExpr and
    not e instanceof MulExpr and
    not e instanceof DivExpr and
    not e instanceof RemExpr and
    not e instanceof BitwiseAndExpr and
    not e instanceof BitwiseOrExpr and
    not e instanceof BitwiseXorExpr and
    not e instanceof LShiftExpr and
    not e instanceof RShiftExpr and
    not e instanceof ConditionalExpr and
    not e instanceof RefExpr and
    not e instanceof LocalVariableDeclAndInitExpr and
    not e instanceof SwitchCaseExpr and
    not e instanceof CastExpr and
    not e instanceof SwitchExpr and
    not e instanceof NullCoalescingExpr
  }

  /** Gets a possible sign for `e` from the signs of its child nodes. */
  Sign specificSubExprSign(Expr e) {
    // The expression types that are handled here should be excluded in `unknownIntegerAccess`.
    // Keep them in sync.
    result = exprSign(e.(AssignExpr).getRValue())
    or
    result = exprSign(e.(AssignOperation).getExpandedAssignment())
    or
    result = exprSign(e.(UnaryPlusExpr).getOperand())
    or
    result = exprSign(e.(PostIncrExpr).getOperand())
    or
    result = exprSign(e.(PostDecrExpr).getOperand())
    or
    result = exprSign(e.(PreIncrExpr).getOperand()).inc()
    or
    result = exprSign(e.(PreDecrExpr).getOperand()).dec()
    or
    result = exprSign(e.(UnaryMinusExpr).getOperand()).neg()
    or
    result = exprSign(e.(ComplementExpr).getOperand()).bitnot()
    or
    e =
      any(DivExpr div |
        result = exprSign(div.getLeftOperand()) and
        result != TZero() and
        div.getRightOperand().(RealLiteral).getValue().toFloat() = 0
      )
    or
    exists(Sign s1, Sign s2 | binaryOpSigns(e, s1, s2) |
      e instanceof AddExpr and result = s1.add(s2)
      or
      e instanceof SubExpr and result = s1.add(s2.neg())
      or
      e instanceof MulExpr and result = s1.mul(s2)
      or
      e instanceof DivExpr and result = s1.div(s2)
      or
      e instanceof RemExpr and result = s1.rem(s2)
      or
      e instanceof BitwiseAndExpr and result = s1.bitand(s2)
      or
      e instanceof BitwiseOrExpr and result = s1.bitor(s2)
      or
      e instanceof BitwiseXorExpr and result = s1.bitxor(s2)
      or
      e instanceof LShiftExpr and result = s1.lshift(s2)
      or
      e instanceof RShiftExpr and result = s1.rshift(s2)
    )
    or
    result = exprSign(e.(ConditionalExpr).getAChild())
    or
    result = exprSign(e.(NullCoalescingExpr).getAChild())
    or
    result = exprSign(e.(SwitchExpr).getACase().getBody())
    or
    result = exprSign(e.(CastExpr).getExpr())
    or
    result = exprSign(e.(SwitchCaseExpr).getBody())
    or
    result = exprSign(e.(LocalVariableDeclAndInitExpr).getInitializer())
    or
    result = exprSign(e.(RefExpr).getExpr())
  }

  private Sign binaryOpLhsSign(BinaryOperation e) { result = exprSign(e.getLeftOperand()) }

  private Sign binaryOpRhsSign(BinaryOperation e) { result = exprSign(e.getRightOperand()) }

  pragma[noinline]
  private predicate binaryOpSigns(Expr e, Sign lhs, Sign rhs) {
    lhs = binaryOpLhsSign(e) and
    rhs = binaryOpRhsSign(e)
  }

  Expr getARead(Ssa::Definition v) { result = v.getARead() }

  Field getField(FieldAccess fa) { result = fa.getTarget() }

  Expr getAnExpression(SsaReadPositionBlock bb) { result = bb.getBlock().getANode().getElement() }

  Guard getComparisonGuard(ComparisonExpr ce) { result = ce.getExpr() }

  /**
   * Holds if `guard` controls the position `controlled` with the value `testIsTrue`.
   */
  predicate guardControlsSsaRead(Guard guard, SsaReadPosition controlled, boolean testIsTrue) {
    exists(BooleanValue b | b.getValue() = testIsTrue |
      guard.controlsBasicBlock(controlled.(SsaReadPositionBlock).getBlock(), b)
    )
  }

  /** A relational comparison */
  class ComparisonExpr extends ComparisonTest {
    private boolean strict;

    ComparisonExpr() {
      this.getComparisonKind() =
        any(ComparisonKind ck |
          ck.isLessThan() and strict = true
          or
          ck.isLessThanEquals() and
          strict = false
        )
    }

    /**
     * Gets the operand on the "greater" (or "greater-or-equal") side
     * of this relational expression, that is, the side that is larger
     * if the overall expression evaluates to `true`; for example on
     * `x <= 20` this is the `20`, and on `y > 0` it is `y`.
     */
    Expr getGreaterOperand() { result = this.getSecondArgument() }

    /**
     * Gets the operand on the "lesser" (or "lesser-or-equal") side
     * of this relational expression, that is, the side that is smaller
     * if the overall expression evaluates to `true`; for example on
     * `x <= 20` this is `x`, and on `y > 0` it is the `0`.
     */
    Expr getLesserOperand() { result = this.getFirstArgument() }

    /** Holds if this comparison is strict, i.e. `<` or `>`. */
    predicate isStrict() { strict = true }
  }
}
