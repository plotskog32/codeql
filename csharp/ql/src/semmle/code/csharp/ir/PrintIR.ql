/**
 * @name Print IR
 * @description Outputs a representation of the IR graph
 * @id charp/print-ir
 * @kind graph
 */

import implementation.raw.PrintIR
