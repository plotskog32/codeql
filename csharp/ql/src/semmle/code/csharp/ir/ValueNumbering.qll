import implementation.raw.gvn.ValueNumbering
