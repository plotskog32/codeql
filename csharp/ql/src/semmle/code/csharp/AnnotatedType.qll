/**
 * Provides predicates and classes for representing annotations on types, such as nullability information.
 *
 * Use an element's `getType()` predicate to access the standard type of the element, without regard
 * for annotations on the type. Use the corresponding `getAnnotatedType()` predicate to
 * get the annotated type, that includes this additional information.
 */

import csharp

private module Annotations {
  newtype TAnnotation =
    TNotNullableRefType() or
    TNullableRefType() or
    TReadonlyRefType() or
    TRefType() or
    TOutType()

  /** An annotation on a type. */
  class TypeAnnotation extends TAnnotation {
    /** Holds if this string should prefix the type string. */
    predicate isPrefix() { none() }

    /** Holds if this string should suffix the type string. */
    predicate isSuffix() { none() }

    /** Gets the bit position in the bit-field. */
    abstract int getBit();

    /** Gets a string representation of this type annotation. */
    abstract string toString();
  }

  /** An annotation indicating that the type is not nullable. */
  class NonNullableRefType extends TypeAnnotation, TNotNullableRefType {
    override predicate isSuffix() { any() }

    override string toString() { result = "!" }

    override int getBit() { result = 2 }
  }

  /** An annotation indicating that the type is a nullable reference type. */
  class NullableRefType extends TypeAnnotation, TNullableRefType {
    override predicate isSuffix() { any() }

    override string toString() { result = "?" }

    override int getBit() { result = 3 }
  }

  /** An annotation indicating that the type is a readonly reference. */
  class ReadonlyRefType extends TypeAnnotation, TReadonlyRefType {
    override predicate isPrefix() { any() }

    override string toString() { result = "readonly " }

    override int getBit() { result = 4 }
  }

  /** An annotation indicating that the variable or return is by `ref`. */
  class RefTypeAnnotation extends TypeAnnotation, TRefType {
    override predicate isPrefix() { any() }

    override string toString() { result = "ref " }

    override int getBit() { result = 5 }
  }

  /** An annotation indicating that the parameter is `out`. */
  class OutType extends TypeAnnotation, TOutType {
    override predicate isPrefix() { any() }

    override string toString() { result = "out " }

    override int getBit() { result = 6 }
  }

  newtype TAnnotations =
    TAnnotationFlags(int flags, Nullability n) {
      exists(Element e | flags = getElementTypeFlags(e) and n = getElementNullability(e))
      or
      flags = 0 // n is unbound
    }

  class TypeAnnotations extends TAnnotations {
    int flags;

    Nullability nullability;

    TypeAnnotations() { this = TAnnotationFlags(flags, nullability) }

    /** Gets the flags (as a bitset) of this type annotation. */
    int getFlags() { result = flags }

    /** Gets the nullability of this type annotation. */
    Nullability getNullability() { result = nullability }

    /**
     * Gets the `i`th "child" of this type annotation.
     * This is used to represent structured datatypes, where the structure
     * of the type annotation mirrors the structure of the annotated type.
     */
    bindingset[i]
    TypeAnnotations getChild(int i) {
      result.getFlags() = 0 and result.getNullability() = this.getNullability().getMember(i)
    }

    /** Gets text to be displayed before the type. */
    string getTypePrefix() {
      result = concat(TypeAnnotation a |
          a = this.getAnAnnotation() and a.isPrefix()
        |
          a.toString(), "" order by a.getBit()
        )
    }

    /** Gets text to be displayed after the type. */
    string getTypeSuffix() {
      result = concat(TypeAnnotation a |
          a = this.getAnAnnotation() and a.isSuffix()
        |
          a.toString(), "" order by a.getBit()
        )
    }

    /** Gets a textual representation of this type annotation. */
    string toString() { result = getTypePrefix() + getTypeSuffix() }

    private predicate isSet(int bit) {
      isBit(bit) and
      exists(int mask | mask = getBitMask(bit) | flags.bitAnd(mask) = mask)
    }

    /** Gets an annotation in this set of annotations. */
    TypeAnnotation getAnAnnotation() {
      isSet(result.getBit())
      or
      nullability instanceof AnnotatedNullability and result instanceof NullableRefType
      or
      nullability instanceof NotAnnotatedNullability and result instanceof NonNullableRefType
    }
  }

  /**
   * A structured type annotation representing type nullability.
   * For example, `IDictionary<string!,string?>?` has nullability `<!,?>?`.
   */
  class Nullability extends @nullability {
    string toString() { result = getMemberString() + getSelfNullability() }

    language[monotonicAggregates]
    private string getMemberString() {
      if nullability_member(this, _, _)
      then
        result = "<" +
            concat(int i, Nullability child |
              nullability_member(this, i, child)
            |
              child.toString(), "," order by i
            ) + ">"
      else result = ""
    }

    /**
     * Gets the `i`th member of this annotation.
     * Returns `this` if the nullability is not explicitly
     * stored in the database, since many type annotations will have consistent
     * nullability.
     */
    bindingset[i]
    Nullability getMember(int i) {
      if nullability_member(this, i, _) then nullability_member(this, i, result) else result = this
    }

    /** Gets a string representing the nullability. */
    abstract string getSelfNullability();
  }

  /**
   * A type that is "oblivious", either because nullability is not
   * applicable, because the code was not compiled in a nullable context, or
   * because the C# language version is less than 8.
   */
  class ObliviousNullability extends Nullability, @oblivious {
    override string getSelfNullability() { result = "_" }
  }

  /**
   * A type that is "fully" oblivious. The type itself is oblivious
   * and all type arguments are oblivious.
   */
  class NoNullability extends ObliviousNullability {
    NoNullability() { not nullability_member(this, _, _) }
  }

  /** A type with annotated nullablity, `?`. */
  class AnnotatedNullability extends Nullability, @annotated {
    override string getSelfNullability() { result = "?" }
  }

  /**
   * A ref type not annotated with `?` in a nullable context.
   */
  class NotAnnotatedNullability extends Nullability, @not_annotated {
    override string getSelfNullability() { result = "!" }
  }

  /** Holds if the type annotations `annotations` apply to type `type` on element `element`. */
  predicate elementTypeAnnotations(
    @has_type_annotation element, Type type, TypeAnnotations annotations
  ) {
    annotations = TAnnotationFlags(getElementTypeFlags(element), getElementNullability(element)) and
    (
      type = element.(Assignable).getType()
      or
      type = element.(Callable).getReturnType()
      or
      type = element.(Expr).getType()
      or
      type = element.(DelegateType).getReturnType()
    )
  }
}

/** Holds if `bit` is a valid bit to set in the bitmask. */
private predicate isBit(int bit) { bit = any(Annotations::TypeAnnotation a).getBit() }

private int getBitMask(int bit) {
  isBit(bit) and
  result = 1.bitShiftLeft(bit)
}

private int getElementTypeFlags(@has_type_annotation element) {
  result = sum(int b | type_annotation(element, b) | b)
}

private Annotations::Nullability getTypeParameterNullability(
  TypeParameterConstraints constraints, Type type
) {
  if specific_type_parameter_nullability(constraints, getTypeRef(type), _)
  then specific_type_parameter_nullability(constraints, getTypeRef(type), result)
  else (
    specific_type_parameter_constraints(constraints, type) and
    result instanceof Annotations::NoNullability
  )
}

private Annotations::Nullability getElementNullability(@has_type_annotation element) {
  if type_nullability(element, _)
  then type_nullability(element, result)
  else result instanceof Annotations::NoNullability
}

private newtype TAnnotatedType =
  TAnnotatedTypeNullability(Type type, Annotations::TypeAnnotations annotations) {
    Annotations::elementTypeAnnotations(_, type, annotations)
    or
    exists(AnnotatedConstructedType c, int i |
      type = c.getType().(ConstructedType).getTypeArgument(i) and
      annotations = c.getAnnotations().getChild(i)
    )
    or
    annotations.getFlags() = 0 and
    annotations.getNullability() = getTypeParameterNullability(_, type)
    or
    // All types exist as AnnotatedTypes with NoNullability.
    annotations.getFlags() = 0 and
    annotations.getNullability() instanceof Annotations::NoNullability
    or
    exists(AnnotatedArrayType at |
      type = at.getType().(ArrayType).getElementType() and
      annotations = at.getAnnotations().getChild(0)
    )
  }

/** A type with additional type information. */
class AnnotatedType extends TAnnotatedType {
  Type type;

  Annotations::TypeAnnotations annotations;

  AnnotatedType() { this = TAnnotatedTypeNullability(type, annotations) }

  /** Gets a textual representation of this annotated type. */
  string toString() {
    result = annotations.getTypePrefix() + getUnderlyingType().toStringWithTypes() +
        annotations.getTypeSuffix()
  }

  /** Gets the location of this annotated type. */
  Location getLocation() { result = type.getLocation() }

  /**
   * Gets the unannotated type, for example `string` in `string?`.
   * Note that this might be a nullable value type (`System.Nullable`).
   */
  Type getType() { result = type }

  /**
   * Gets the underlying type, for example `string` in `string?`
   * or `int` in `int?`. This also gets the underlying type of
   * nullable value types (`System.Nullable`).
   */
  final Type getUnderlyingType() {
    if type instanceof NullableType
    then result = type.(NullableType).getUnderlyingType()
    else result = type
  }

  /** Gets the type annotation set of this annotated type. */
  Annotations::TypeAnnotations getAnnotations() { result = annotations }

  /** Gets a type annotation of this annotated type. */
  private Annotations::TypeAnnotation getAnAnnotation() {
    result = getAnnotations().getAnAnnotation()
  }

  /** Holds if the type is a non-nullable reference, for example, `string` in a nullable-enabled context. */
  predicate isNonNullableRefType() {
    this.getAnAnnotation() instanceof Annotations::NonNullableRefType
  }

  /** Holds if the type is a nullable reference, for example `string?`. */
  predicate isNullableRefType() { this.getAnAnnotation() instanceof Annotations::NullableRefType }

  /** Holds if the type is a `ref`, for example the return type of `ref int F()`. */
  predicate isRef() { this.getAnAnnotation() instanceof Annotations::RefTypeAnnotation }

  /** Holds if the type is a `ref readonly`, for example the return type of `ref readonly int F()`. */
  predicate isReadonlyRef() { this.getAnAnnotation() instanceof Annotations::ReadonlyRefType }

  /** Holds if the type is an `out`, for example parameter `p` in `void F(out int p)`. */
  predicate isOut() { this.getAnAnnotation() instanceof Annotations::OutType }

  /** Holds if this annotated type applies to element `e`. */
  predicate appliesTo(Element e) { Annotations::elementTypeAnnotations(e, type, annotations) }

  /** Holds if this annotated type is the type argument 'i' of constructed generic 'g'. */
  predicate appliesToTypeArgument(ConstructedGeneric g, int i) {
    this.getAnnotations().getFlags() = 0 and
    this.getAnnotations().getNullability() = getElementNullability(g).getMember(i) and
    this.getType() = g.getTypeArgument(i)
  }

  /** Holds if this annotated type applies to type parameter constraints `constraints`. */
  predicate appliesToTypeConstraint(TypeParameterConstraints constraints) {
    this.getAnnotations().getFlags() = 0 and
    this.getAnnotations().getNullability() = getTypeParameterNullability(constraints, type)
  }
}

/** An array type with additional type information. */
class AnnotatedArrayType extends AnnotatedType {
  override ArrayType type;

  /** Gets the annotated element type of this array, for example `int?` in `int?[]`. */
  final AnnotatedType getElementType() {
    result.getType() = type.getElementType() and
    result.getAnnotations() = this.getAnnotations().getChild(0)
  }

  private string getDimensionString(AnnotatedType elementType) {
    exists(AnnotatedType et, string res |
      et = getElementType() and
      res = "[" + type.getRankString(0) + "]" and
      if et.getUnderlyingType() instanceof ArrayType and not et.isNullableRefType()
      then result = res + et.(AnnotatedArrayType).getDimensionString(elementType)
      else (
        result = res and elementType = et
      )
    )
  }

  override string toString() {
    exists(AnnotatedType elementType |
      result = annotations.getTypePrefix() + elementType.toString() +
          this.getDimensionString(elementType) + annotations.getTypeSuffix()
    )
  }
}

/** A constructed type with additional type information. */
class AnnotatedConstructedType extends AnnotatedType {
  override ConstructedType type;

  /** Gets the `i`th type argument of this constructed type. */
  AnnotatedType getTypeArgument(int i) {
    result.getType() = type.getTypeArgument(i) and
    result.getAnnotations() = this.getAnnotations().getChild(i)
  }

  override string toString() {
    result = annotations.getTypePrefix() + type.getUnboundGeneric().getNameWithoutBrackets() + "<" +
        this.getTypeArgumentsString() + ">" + annotations.getTypeSuffix()
  }

  language[monotonicAggregates]
  private string getTypeArgumentsString() {
    result = concat(int i |
        exists(this.getTypeArgument(i))
      |
        this.getTypeArgument(i).toString(), ", " order by i
      )
  }
}
