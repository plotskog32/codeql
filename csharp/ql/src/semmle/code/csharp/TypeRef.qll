/**
 * INTERNAL: Do not use.
 * Provides support for type-references.
 */

import csharp

/**
 * INTERNAL: Do not use.
 * Gets a type reference for a given type `type`.
 * This is used for extensionals that can be supplied
 * as either type references or types.
 */
@type_or_ref getTypeRef(@type type) {
  result = type
  or
  type = result.(TypeRefs::TypeRef).getCanonicalType()
}

private module TypeRefs {
  /**
   * A typeref is a reference to a type in some assembly.
   * Often, a type can be present in multiple assemblies.
   */
  class TypeRef extends @typeref {
    string getName() { typerefs(this, result) }

    string toString() { result = getName() }

    Type getAType() { typeref_type(this, result) }

    /**
     * Gets the "canonical type" represented by a typeref.
     * This is because a typeref can reference multiple types, or a type with the same
     * fully qualified name can exist in multiple assemblies.
     * The canonical type is an arbitrarily chosen type from the list of candidate types.
     */
    Type getCanonicalType() {
      result = this.getAType() and
      isCanonicalType(result)
    }
  }

  /** Gets a location of a type. */
  private Location typeLocation(Type t) {
    type_location(t, result)
    or
    exists(Type decl | constructed_generic(decl, t) and result = typeLocation(decl))
  }

  /** Gets the "canonical location" for a type. A type has only one canonical location. */
  private Location canonicalTypeLocation(Type t) {
    result = typeLocation(t) and
    not locationIsBetter(result, typeLocation(t))
  }

  pragma[inline]
  private predicate locationIsBetter(Location typeLocation, Location betterLocation) {
    typeLocation.(Assembly).getFullName() < betterLocation.(Assembly).getFullName()
    or
    typeLocation instanceof Assembly and betterLocation instanceof SourceLocation
    or
    typeLocation.(SourceLocation).getFile().getAbsolutePath() <
      betterLocation.(SourceLocation).getFile().getAbsolutePath()
  }

  private predicate isCanonicalType(Type type) {
    not exists(TypeRef tr |
      type = tr.getAType() and
      locationIsBetter(canonicalTypeLocation(type), canonicalTypeLocation(tr.getAType()))
    )
  }
}
