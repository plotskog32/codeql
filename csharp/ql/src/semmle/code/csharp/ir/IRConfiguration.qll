import csharp

private newtype TIRConfiguration = MkIRConfiguration()

/**
 * The query can extend this class to control which functions have IR generated for them.
 */
class IRConfiguration extends TIRConfiguration {
  string toString() {
    result = "IRConfiguration"
  }

  /**
   * Holds if IR should be created for method `method`. By default, holds for all method.
   */
  predicate shouldCreateIRForFunction(Callable callable) {
    any()
  }
}
