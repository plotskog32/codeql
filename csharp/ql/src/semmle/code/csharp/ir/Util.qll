import csharp

class Locatable extends Element {
	
}

class ArrayInitWithMod extends ArrayInitializer {
	predicate isInitialized(int entry) {
		entry in [0..this.getNumberOfElements() - 1]
	}
	
	predicate isValueInitialized(int entry) {
		isInitialized(entry) and
		not exists()
	}
}