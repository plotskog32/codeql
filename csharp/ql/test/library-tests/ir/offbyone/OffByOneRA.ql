import csharp
import semmle.code.csharp.ir.IR
import semmle.code.csharp.ir.rangeanalysis.RangeAnalysis

/**
 * Holds if the index expression of `aa` is less than or equal to the array length plus `k`.
 */
predicate boundedArrayAccess(ElementAccess aa, int k) {
  exists(Instruction index, Instruction usage, Bound b, int delta |
    (
      // indexer access
      usage.(CallInstruction).getAST() = aa
      or
      // array access
      usage.(PointerAddInstruction).getAST() = aa
    ) and
    usage.getAnOperand().getDef() = index and
    boundedInstruction(index, b, delta, true, _)
  |
    exists(PropertyAccess pa |
      k = delta and
      b.getInstruction().getAST() = pa and
      pa.getProperty().getName() = "Length" and
      pa.(QualifiableExpr).getQualifier().(VariableAccess).getTarget() = aa
            .getQualifier()
            .(VariableAccess)
            .getTarget()
    )
    or
    exists(ArrayCreation ac |
      ac.getParent().(LocalVariableDeclExpr).getVariable() = aa
            .getQualifier()
            .(VariableAccess)
            .getTarget() and
      b instanceof ZeroBound and
      if exists(ac.getLengthArgument(0))
      then k = delta - ac.getLengthArgument(0).getValue().toInt()
      else
        if exists(ac.getInitializer())
        then k = delta - ac.getInitializer().getNumberOfElements()
        else none()
    )
  )
}

/**
 * Holds if the index expression is less than or equal to the array length plus `k`,
 * but not necessarily less than or equal to the array length plus `k-1`.
 */
predicate bestArrayAccessBound(ElementAccess aa, int k) {
  k = min(int k0 | boundedArrayAccess(aa, k0))
}

from ElementAccess aa, int k, string msg, string add
where
  bestArrayAccessBound(aa, k) and
  (if k = 0 then add = "" else add = " + " + k) and
  (
    if k >= 0
    then
      msg = "This array access might be out of bounds, as the index might be equal to the array length"
          + add
    else msg = "This array access is ok, the index is at most the lenth of the array " + add
  )
select aa.getEnclosingCallable(), aa, msg
