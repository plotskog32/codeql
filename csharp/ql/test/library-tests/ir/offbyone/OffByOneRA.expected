| test.cs:10:17:10:21 | test1 | test.cs:17:41:17:51 | access to array element | This array access is ok, the index is at most the lenth of the array  + -1 |
| test.cs:21:17:21:21 | test2 | test.cs:33:41:33:51 | access to array element | This array access might be out of bounds, as the index might be equal to the array length |
| test.cs:37:17:37:21 | test3 | test.cs:50:41:50:51 | access to array element | This array access is ok, the index is at most the lenth of the array  + -1 |
| test.cs:54:17:54:21 | test4 | test.cs:64:41:64:51 | access to array element | This array access might be out of bounds, as the index might be equal to the array length |
| test.cs:68:17:68:21 | test5 | test.cs:74:22:74:27 | access to indexer | This array access might be out of bounds, as the index might be equal to the array length |
| test.cs:78:17:78:21 | test6 | test.cs:87:41:87:55 | access to array element | This array access might be out of bounds, as the index might be equal to the array length |
| test.cs:91:17:91:21 | test7 | test.cs:101:41:101:50 | access to array element | This array access might be out of bounds, as the index might be equal to the array length + 1 |
